functions/src/index.ts

```diff
@@ -57,6 +70,94 @@ export const finishGame = functions.https.onCall(async (data, context) => {
     status: "done",
     endedAt: finalTime,
   });
+
+  // Update ratings of players involved based on an extension of the Elo system.
+
+  // Retrieve userIds of players in the game.
+  const playersSnap = await admin
+    .database()
+    .ref(`games/${gameId}/users`)
+    .once("value");
+  const players: string[] = [];
+  playersSnap.forEach(function (childSnap) {
+    if (childSnap.key !== null) {
+      players.push(childSnap.key);
+    }
+  });
+
+  // Count total number of sets in the game.
+  let setCount = 0;
+  for (const score of scores.values()) {
+    setCount += score;
+  }
+
+  // Add scores for players without a single set.
+  for (const player of players) {
+    if (!scores.has(player)) {
+      scores.set(player, 0);
+    }
+  }
+
+  // Retrieve old ratings from the database.
+  const ratings = new Map<string, number>();
+  for (const player of players) {
+    const ratingSnap = await admin
+      .database()
+      .ref(`users/${player}/ratings/${gameMode}`)
+      .once("value");
+    const rating = ratingSnap.exists() ? ratingSnap.val() : BASE_RATING;
+    ratings.set(player, rating);
+  }
+
+  // Translate ratings to exponential format.
+  const expRatings = new Map<string, number>();
+  for (const player of players) {
+    expRatings.set(
+      player,
+      Math.pow(10, <number>ratings.get(player) / SCALING_FACTOR)
```

Comment:
Not sure why these coercions to `<number>` are needed everywhere. Is the issue that it could be nullable? If that's the case and it's not compiling, then you should use the non-nullable assertion operator, which I believe is `!`.

Reply:
Ah, that is indeed a much cleaner solution.